Source/WebCore/dom/Subscriber.h

```diff
@@ -0,0 +1,111 @@
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include "AbortController.h"
+#include "ActiveDOMObject.h"
+#include "ScriptExecutionContext.h"
+#include "ScriptWrappable.h"
+#include "SubscriptionObserverCallback.h"
+#include "VoidCallback.h"
+#include <wtf/RefCounted.h>
+
+namespace WebCore {
+
+class Subscriber final : public ActiveDOMObject, public ScriptWrappable, public RefCounted<Subscriber> {
+    WTF_MAKE_ISO_ALLOCATED(Subscriber);
+
+public:
+    void next(JSC::JSValue);
+    void complete();
+    void error(JSC::JSValue);
+    void addTeardown(Ref<VoidCallback>);
+
+    bool active() { return m_active; }
+    AbortSignal& signal() { return m_abortController->signal(); }
+
+    static Ref<Subscriber> create(ScriptExecutionContext&);
+
+    static Ref<Subscriber> create(ScriptExecutionContext&, RefPtr<SubscriptionObserverCallback> next);
+
+    static Ref<Subscriber> create(ScriptExecutionContext&, RefPtr<SubscriptionObserverCallback> next,
+        RefPtr<SubscriptionObserverCallback> error,
+        RefPtr<VoidCallback>);
+
+    explicit Subscriber(ScriptExecutionContext&, RefPtr<SubscriptionObserverCallback> next,
+        RefPtr<SubscriptionObserverCallback> error,
+        RefPtr<VoidCallback>);
+
+    void followSignal(AbortSignal&);
+
+    // JSCustomMarkFunction; for JSSubscriberCustom
+    SubscriptionObserverCallback* nextCallbackConcurrently() { return m_next.get(); }
+    SubscriptionObserverCallback* errorCallbackConcurrently() { return m_error.get(); }
+    VoidCallback* completeCallbackConcurrently() { return m_complete.get(); }
+    std::vector<VoidCallback*> teardownCallbacksConcurrently()
+    {
+        // Copy into a vector to unpack Ref<>s
+        std::vector<VoidCallback*> callbacks;
+        callbacks.reserve(m_teardowns.size());
+        for (auto callback : m_teardowns)
+            callbacks.push_back(callback.ptr());
+        return callbacks;
+    }
+
+    // ActiveDOMObject
+    void ref() const final { RefCounted::ref(); }
+    void deref() const final { RefCounted::deref(); }
+
+private:
+    bool m_active = true;
+    Ref<AbortController> m_abortController;
+    RefPtr<SubscriptionObserverCallback> m_next;
+    RefPtr<SubscriptionObserverCallback> m_error;
+    RefPtr<VoidCallback> m_complete;
+    ListHashSet<Ref<VoidCallback>> m_teardowns;
```

Comment:
Why a `ListHashSet`? We don't ever look up a callback, only ever appending or iterating through the whole structure. I think a `Vector` would work fine for this.

Reply:
To make it thread safe you can add a `mutable Lock` and then annotate this with a `WTF_GUARDED_BY_LOCK(m_teardownsLock);` so thread safety analysis works.